pkg/recommendation/device.go

```diff
@@ -0,0 +1,235 @@
+package recommendation
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"mayadata.io/cstorpoolauto/types"
+	bdutil "mayadata.io/cstorpoolauto/util/blockdevice"
+)
+
+// nodeCapacityBlockDevice contains key with node name and value with capacityBlockDevice
+type nodeCapacityBlockDevice map[string]capacityBlockDevices
+
+// capacityBlockDevices contains a key with capacity of block device and
+// value with all the block devices of that capacity.
+type capacityBlockDevices map[int64][]bdutil.MetaInfo
+
+type cStorPoolClusterRecommendationRequest types.CStorPoolClusterRecommendationRequest
+
+// NewDeviceRequest returns a device request object after validation.
+func NewDeviceRequest(
+	raidConfig *types.RaidGroupConfig, blockDeviceList *unstructured.UnstructuredList, poolCapacity *resource.Quantity) (
+	*cStorPoolClusterRecommendationRequest, error) {
+
+	if raidConfig == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil raid config")
+	}
+	if blockDeviceList == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil block device list")
+	}
+
+	err := raidConfig.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create capacity recommendation request")
+	}
+
+	return &cStorPoolClusterRecommendationRequest{
+		Spec: types.CStorPoolClusterRecommendationRequestSpec{
+			PoolCapacity:    *poolCapacity,
+			BlockDeviceList: *blockDeviceList,
+			DataConfig:      *raidConfig,
+		},
+	}, nil
+}
+
+// GetRecommendation returns recommended block devices for all nodes and device types.
+func (r cStorPoolClusterRecommendationRequest) GetRecommendation() map[string]types.CStorPoolClusterRecommendation {
```

Comment:
IMO we should not expose methods out of types in this way.
There should be separate structures against whom methods should get exposed.

Reply:
I am not sure how this works. How are we avoiding method name clashes?
Have we done type aliasing to avoid these duplicate method name clash?